benchexec/tools/template.py

```diff
@@ -22,17 +35,386 @@ class UnsupportedFeatureException(benchexec.BenchExecException):
     pass
 
 
-class BaseTool(object):
+class BaseTool2(object, metaclass=ABCMeta):
     """
     This class serves both as a template for tool-info implementations,
     and as an abstract super class for them.
     For writing a new tool info, inherit from this class and override
     the necessary methods (always executable() and name(),
     usually determine_result(), cmdline(), and version(),
     and maybe working_directory() and get_value_from_output(), too).
-    The class for each specific tool need to be named "Tool".
-    For more information, please refer to
-    https://github.com/sosy-lab/benchexec/blob/master/doc/tool-integration.md
+
+    In special circumstances, it can make sense to not inherit from this class.
+    In such cases the tool-info module's class needs to implement all the methods
+    defined here and BaseTool2.register needs to be called with the respective class
+    (cf. documentation of abc.ABCMeta.register) to declare compatibility with BaseTool2.
+    Note that we might add optional methods (with default implementations) to BaseTool2
+    at any time.
+
+    This class is supported since BenchExec 3.2.
+    """
+
+    REQUIRED_PATHS = []
+    """
+    List of path patterns that is used by the default implementation of program_files().
+    Not necessary if this method is overwritten.
+    """
+
+    # Methods that provide general (run-independent) information about the tool
+
+    @abstractmethod
+    def name(self):
+        """
+        Return the name of the tool, formatted for humans.
+        This function should always be overriden.
+        @return a non-empty string
+        """
+        return "UNKOWN"
+
+    @abstractmethod
+    def executable(self):
+        """
+        Find the path to the executable file that will get executed.
+        This method always needs to be overridden,
+        and most implementations will look similar to this one.
+        The path returned should be relative to the current directory.
+        @return a string pointing to an executable file
+        """
+        return util.find_executable("tool")
```

Comment:
Subclasses of `BaseTool2` can call `super().executable()`, which wouldn't make sense.

It may be more safe to replace the method definition with `raise NotImplementedError()`, but I like that an example is given.

Similar reasoning for the other abstract methods of this class.

Reply:
This would be possible, but I guess it is not necessary. Calling `super().executable()` is not something that would be done automatically, and it would also immediately fail and be detected, right?

Reply:
But why go with the less safe way, if there's a more safe (and more expressive) alternative?

The only argument I see is having a 'working' example.
But if you put the code example in the comments, IDE refactoring will still find it in case of method renames or method moves.

Reply:
Hm, my thought was to have
```
raise NotImplementedError()
return util.find_executable("tool")
```
but this looked ugly and confusing.

A comment is maybe a better idea.

Reply:
We could even test the examples with [`doctest`](https://docs.python.org/3.5/library/doctest.html).
Unfortunately, doctest seems to require these ugly `>>>` to recognize code examples..